File: libs/fleet/sdk-bindings/examples/ruby/live_app_controlled.rb
#!/usr/bin/env ruby
require 'json'
require 'net/http'
require 'uri'
require 'cyclops_sdk'

class NetHttpClient < FleetSdk::HttpClient
  def execute(request)
    uri = URI(request.url)
    native = Net::HTTPGenericRequest.new(request.method, !request.body.nil?, request.method != 'HEAD', uri.request_uri, request.headers.to_h { |header| [header.name, header.value] })
    native.body = request.body unless request.body.nil?
    response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https', open_timeout: 30, read_timeout: 60) { |http| http.request(native) }
    FleetSdk::HttpResponse.new(status: response.code.to_i, headers: response.each_header.map { |name, value| FleetSdk::HttpHeader.new(name: name, value: value) }, body: response.body.to_s.b)
  rescue StandardError => error
    raise FleetSdk::HttpError::Transport.new(reason: error.message)
  end
end

def required(name)
  ENV.fetch(name).tap { |value| raise "#{name} is required" if value.empty? }
end

def pool_spec
  FleetSdk::PoolSpec.new(replicas: 1, template: FleetSdk::PoolTemplate.new(runtime: nil, runtime_class_name: nil, node_selector: nil, tolerations: nil, command: nil, container_disk_image: required('CUA_IMAGE'), image_pull_secret: required('CUA_IMAGE_PULL_SECRET'), cpu_cores: 4, memory: '4Gi', firmware: nil, probes: nil, oidc: nil), autoscaling: nil, services: [FleetSdk::SandboxService.new(name: 'mcp', target_port: 3000, protocol: nil)])
end

def initialize_mcp(client, sandbox)
  body = JSON.generate(jsonrpc: '2.0', id: 1, method: 'initialize', params: { protocolVersion: '2025-03-26', capabilities: {}, clientInfo: { name: 'cyclops-uniffi-ruby', version: '0.1.0' } }).b
  deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + 300
  loop do
    response = client.service_request(sandbox, 'mcp', '/mcp', FleetSdk::HttpRequest.new(method: 'POST', url: 'https://ignored.invalid/mcp', headers: [FleetSdk::HttpHeader.new(name: 'accept', value: 'application/json, text/event-stream'), FleetSdk::HttpHeader.new(name: 'content-type', value: 'application/json')], body: body))
    return response.status if response.status.between?(200, 299)
    if [502, 503, 504].include?(response.status) && Process.clock_gettime(Process::CLOCK_MONOTONIC) < deadline
      sleep 5
      next
    end
    raise "MCP initialize failed with HTTP #{response.status}: #{response.body.inspect}"
  end
end

namespace = required('CYCLOPS_NAMESPACE')
credentials = FleetSdk::CyclopsCredentials.new(required('CUA_CLIENT_ID'), required('CUA_CLIENT_SECRET'))
configuration = FleetSdk::CyclopsConfiguration.new(base_url: required('CUA_BASE_URL'), token_url: required('CUA_TOKEN_URL'), credentials: credentials, pool_poll_interval_ms: 5000, pool_poll_limit: 120, claim_poll_interval_ms: 5000, claim_poll_limit: 120)
client = FleetSdk::CyclopsClient.connect(configuration, NetHttpClient.new)
pool = nil
claim = nil
begin
  pool = client.create_pool(FleetSdk::CreatePoolRequest.new(namespace: namespace, spec: pool_spec))
  claim = client.create_claim(FleetSdk::CreateClaimRequest.new(pool: pool, spec: nil))
  sandbox = client.wait_claim(claim)
  status = initialize_mcp(client, sandbox)
  puts JSON.generate(namespace: namespace, pool: pool.metadata.name, claim: claim.metadata.name, sandbox: sandbox.name, mcp_status: status)
ensure
  begin
    client.delete_claim(claim) unless claim.nil?
  ensure
    client.delete_pool(pool) unless pool.nil?
  end
end

File: libs/fleet/sdk-bindings/examples/python/live_app_controlled.py
#!/usr/bin/env python3
import asyncio
import json
import os
import time
import urllib.error
import urllib.request

from cyclops_sdk import (
    CreateClaimRequest,
    CreatePoolRequest,
    CyclopsClient,
    CyclopsConfiguration,
    CyclopsCredentials,
    HttpClient,
    HttpHeader,
    HttpRequest,
    HttpResponse,
    PoolSpec,
    PoolTemplate,
    SandboxService,
)


class UrlLibHttpClient(HttpClient):
    async def execute(self, request: HttpRequest) -> HttpResponse:
        return await asyncio.to_thread(self._execute, request)

    def _execute(self, request: HttpRequest) -> HttpResponse:
        native = urllib.request.Request(
            request.url,
            data=request.body,
            method=request.method,
            headers={header.name: header.value for header in request.headers},
        )
        try:
            with urllib.request.urlopen(native, timeout=60) as response:
                return HttpResponse(
                    status=response.status,
                    headers=[HttpHeader(name=name, value=value) for name, value in response.headers.items()],
                    body=response.read(),
                )
        except urllib.error.HTTPError as error:  # lint-ignore: swallowed-exception
            return HttpResponse(
                status=error.code,
                headers=[HttpHeader(name=name, value=value) for name, value in error.headers.items()],
                body=error.read(),
            )


def pool_spec(image: str, image_pull_secret: str) -> PoolSpec:
    return PoolSpec(
        replicas=1,
        template=PoolTemplate(
            runtime=None,
            runtime_class_name=None,
            node_selector=None,
            tolerations=None,
            command=None,
            container_disk_image=image,
            image_pull_secret=image_pull_secret,
            cpu_cores=4,
            memory="4Gi",
            firmware=None,
            probes=None,
            oidc=None,
        ),
        autoscaling=None,
        services=[SandboxService(name="mcp", target_port=3000, protocol=None)],
    )


async def initialize_mcp(client: CyclopsClient, sandbox) -> int:
    body = json.dumps(
        {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {
                "protocolVersion": "2025-03-26",
                "capabilities": {},
                "clientInfo": {"name": "cyclops-uniffi-python", "version": "0.1.0"},
            },
        },
        separators=(",", ":"),
    ).encode()
    deadline = time.monotonic() + 300
    while True:
        response = await client.service_request(
            sandbox,
            "mcp",
            "/mcp",
            HttpRequest(
                method="POST",
                url="https://ignored.invalid/mcp",
                headers=[
                    HttpHeader(name="accept", value="application/json, text/event-stream"),
                    HttpHeader(name="content-type", value="application/json"),
                ],
                body=body,
            ),
        )
        if 200 <= response.status < 300:
            return response.status
        if response.status in (502, 503, 504) and time.monotonic() < deadline:
            await asyncio.sleep(5)
            continue
        raise RuntimeError(f"MCP initialize failed with HTTP {response.status}: {response.body!r}")


async def main() -> None:
    client_id = os.environ["CUA_CLIENT_ID"]
    client_secret = os.environ["CUA_CLIENT_SECRET"]
    base_url = os.environ["CUA_BASE_URL"]
    token_url = os.environ["CUA_TOKEN_URL"]
    namespace = os.environ["CYCLOPS_NAMESPACE"]
    image_pull_secret = os.environ["CUA_IMAGE_PULL_SECRET"]
    image = os.environ["CUA_IMAGE"]

    client = CyclopsClient.connect(
        CyclopsConfiguration(
            base_url=base_url,
            token_url=token_url,
            credentials=CyclopsCredentials(client_id, client_secret),
            pool_poll_interval_ms=5000,
            pool_poll_limit=120,
            claim_poll_interval_ms=5000,
            claim_poll_limit=120,
        ),
        UrlLibHttpClient(),
    )
    pool = None
    claim = None
    try:
        pool = await client.create_pool(
            CreatePoolRequest(namespace=namespace, spec=pool_spec(image, image_pull_secret))
        )
        claim = await client.create_claim(CreateClaimRequest(pool=pool, spec=None))
        sandbox = await client.wait_claim(claim)
        status = await initialize_mcp(client, sandbox)
        print(
            json.dumps(
                {
                    "namespace": namespace,
                    "pool": pool.metadata.name,
                    "claim": claim.metadata.name,
                    "sandbox": sandbox.name,
                    "mcp_status": status,
                },
                sort_keys=True,
            )
        )
    finally:
        if claim is not None:
            await client.delete_claim(claim)
        if pool is not None:
            await client.delete_pool(pool)


asyncio.run(main())
